File: updateversionlibrary/src/main/java/com/zhangteng/updateversionlibrary/callback/VersionInfoCallback.java
package com.zhangteng.updateversionlibrary.callback;

import android.annotation.SuppressLint;
import android.app.DialogFragment;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.text.TextUtils;
import android.util.Log;
import android.widget.Toast;

import androidx.fragment.app.FragmentManager;

import com.zhangteng.updateversionlibrary.R;
import com.zhangteng.updateversionlibrary.UpdateVersion;
import com.zhangteng.updateversionlibrary.dialog.UpdateDialogFragment;
import com.zhangteng.updateversionlibrary.entity.VersionEntity;
import com.zhangteng.updateversionlibrary.http.HttpClient;
import com.zhangteng.updateversionlibrary.utils.NetWorkUtils;

import java.io.InputStream;

/**
 * @author swing 2018/5/14
 */
public class VersionInfoCallback {
    @SuppressLint("StaticFieldLeak")
    private static Context mContext;
    private VersionEntity versionEntity;
    private FragmentManager mFragmentManager;
    private HttpClient httpClient;

    /**
     * 开始获取版本信息前的准备工作
     */
    public void onPreExecute(Context context, FragmentManager fragmentManager, HttpClient httpClient) {
        mContext = context;
        this.mFragmentManager = fragmentManager;
        this.httpClient = httpClient;
    }

    /**
     * 从背景任务中获取版本信息
     */
    public void doInBackground(VersionEntity versionEntity) {
        this.versionEntity = versionEntity;
    }

    /**
     * 请求完成后进行下载请求
     */
    public void onPostExecute() {
        if (mContext != null && versionEntity != null) {
            Log.i("auto update", "versionEntity versioncode: " + versionEntity.getVersionNo() + " package versioncode: " + getPackageInfo().versionCode);
            if (versionEntity.getVersionCode() > getPackageInfo().versionCode) {
                if (versionEntity.getForceUpdate() != 0) {
                    UpdateVersion.setIsAutoInstall(true);
                    UpdateVersion.setIsProgressDialogShow(true);
                    NetWorkUtils netWorkUtils = new NetWorkUtils(mContext);
                    int type = netWorkUtils.getNetType();
                    if (type != 1) {
                        showUpdateUICustom(versionEntity);
                    } else {
                        httpClient.downloadApk(versionEntity, new DownloadCallback());
                    }
                } else {
                    if (UpdateVersion.isUpdateDialogShow()) {
                        showUpdateUICustom(versionEntity);
                    }
                }
            } else {
                if (UpdateVersion.isHintVersion()) {
                    Toast.makeText(mContext, mContext == null ? "当前已是最新版" : mContext.getString(R.string.version_hint), Toast.LENGTH_LONG).show();
                }
            }
        } else {
            if (UpdateVersion.isHintVersion()) {
                Toast.makeText(mContext, mContext == null ? "当前已是最新版" : mContext.getString(R.string.version_hint), Toast.LENGTH_LONG).show();
            }
        }
    }

    /**
     * 获取当前app版本
     */
    private PackageInfo getPackageInfo() {
        PackageInfo pinfo = null;
        if (mContext != null) {
            try {
                pinfo = mContext.getPackageManager().getPackageInfo(
                        mContext.getPackageName(), 0);
            } catch (PackageManager.NameNotFoundException e) {
                e.printStackTrace();
            }
        }
        return pinfo;
    }

    /**
     * 更新提示
     */
    @SuppressLint("WrongConstant")
    private void showUpdateUICustom(final VersionEntity versionEntity) {
        final UpdateDialogFragment dialogFragment = new UpdateDialogFragment();
        dialogFragment.setStyle(DialogFragment.STYLE_NO_TITLE, 0);
        dialogFragment.setTitle(mContext == null ? "发现新版本%s" : String.format(mContext.getString(R.string.version_title), "\n" + versionEntity.getVersionNo()));
        dialogFragment.setContentTitleText(mContext == null ? "更新的内容" : TextUtils.isEmpty(versionEntity.getTitle()) ? mContext.getString(R.string.version_content_title) : versionEntity.getTitle());
        dialogFragment.setContentText(mContext == null ? "%s" : String.format(mContext.getString(R.string.version_content), versionEntity.getUpdateDesc()));
        dialogFragment.setNegativeBtn(mContext == null ? "暂不" : mContext.getString(R.string.version_cancel), null);
        dialogFragment.setPositiveBtn(mContext == null ? "立即更新" : mContext.getString(R.string.version_confirm), () -> {
            NetWorkUtils netWorkUtils = new NetWorkUtils(mContext);
            int type = netWorkUtils.getNetType();
            if (type != 1) {
                showNetCustomDialog(versionEntity);
            } else {
                if (!UpdateVersion.isUpdateDownloadWithBrowser()) {
                    httpClient.downloadApk(versionEntity, new DownloadCallback());
                } else {
                    Intent i = new Intent(Intent.ACTION_VIEW, Uri.parse(versionEntity.getUrl()));
                    mContext.startActivity(i);
                }
            }
        });
        try {
            dialogFragment.show(mFragmentManager, "");
        } catch (IllegalStateException e) {
            Log.e("UpdateDialogFragment", e.getMessage());
        }
    }

    /**
     * 手机网络dialog
     */
    @SuppressLint("WrongConstant")
    private void showNetCustomDialog(final VersionEntity versionEntity) {
        final UpdateDialogFragment dialogFragment = new UpdateDialogFragment();
        dialogFragment.setStyle(DialogFragment.STYLE_NO_TITLE, 0);
        dialogFragment.setContentText(mContext == null ? "当前在无WIFI的情况下下载，请确定是否使用流量继续下载。" : mContext.getString(R.string.no_wifi_hint));
        dialogFragment.setNetHint(true);
        dialogFragment.setNegativeBtn(mContext == null ? "取消" : mContext.getString(R.string.no_wifi_hint_cancel), null);
        dialogFragment.setPositiveBtn(mContext == null ? "继续下载" : mContext.getString(R.string.no_wifi_hint_confirm), () -> {
            if (!UpdateVersion.isUpdateDownloadWithBrowser()) {
                httpClient.downloadApk(versionEntity, new DownloadCallback());
            } else {
                Intent i = new Intent(Intent.ACTION_VIEW, Uri.parse(versionEntity.getUrl()));
                mContext.startActivity(i);
            }
        });
        try {
            dialogFragment.show(mFragmentManager, "");
        } catch (IllegalStateException e) {
            Log.e("UpdateDialogFragment", e.getMessage());
        }
    }

    public static InputStream nativeAssertGet(String URL) {
        InputStream inputStream = null;
        try {
            inputStream = mContext.getAssets().open(URL);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return inputStream;
    }
}

File: updateversion/src/main/java/com/zhangteng/updateversion/callback/VersionInfoCallback.kt
package com.zhangteng.updateversion.callback

import android.annotation.SuppressLint
import android.content.Context
import android.content.ContextWrapper
import android.content.Intent
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.text.TextUtils
import android.util.Log
import android.widget.Toast
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentActivity
import com.zhangteng.updateversion.R
import com.zhangteng.updateversion.UpdateVersion
import com.zhangteng.updateversion.dialog.UpdateDialogFragment
import com.zhangteng.updateversion.entity.VersionEntity
import com.zhangteng.updateversion.http.HttpClient
import com.zhangteng.utils.NetType
import com.zhangteng.utils.getConnectedType
import com.zhangteng.utils.i
import java.io.InputStream

/**
 * @author swing 2018/5/14
 */
open class VersionInfoCallback {
    private var versionEntity: VersionEntity? = null
    private var httpClient: HttpClient? = null

    /**
     * 开始获取版本信息前的准备工作
     */
    fun onPreExecute(
        context: Context?,
        httpClient: HttpClient?
    ) {
        mContext = context
        this.httpClient = httpClient
    }

    /**
     * 从背景任务中获取版本信息
     */
    fun doInBackground(versionEntity: VersionEntity?) {
        this.versionEntity = versionEntity
    }

    /**
     * 请求完成后进行下载请求
     */
    @Suppress("DEPRECATION")
    fun onPostExecute() {
        if (mContext != null && versionEntity != null) {
            val versionCode: Long =
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                    packageInfo!!.longVersionCode
                } else {
                    packageInfo!!.versionCode.toLong()
                }
            "versionEntity versioncode: ${versionEntity!!.versionNo} package versioncode: $versionCode".i(
                "auto update"
            )
            if (versionEntity!!.versionCode > versionCode) {
                if (versionEntity!!.forceUpdate != 0) {
                    //强制更新使用进度弹窗并自动安装
                    UpdateVersion.isAutoInstall = true
                    UpdateVersion.isProgressDialogShow = true
                    UpdateVersion.isNotificationShow = false
                    val type = mContext.getConnectedType()
                    if (type != NetType.Wifi && UpdateVersion.isNetCustomDialogShow) {
                        showUpdateUICustom(versionEntity!!)
                    } else {
                        httpClient?.downloadApk(versionEntity, DownloadCallback())
                    }
                } else {
                    if (UpdateVersion.isUpdateDialogShow) {
                        showUpdateUICustom(versionEntity!!)
                    } else {
                        val type = mContext.getConnectedType()
                        if (type != NetType.Wifi && UpdateVersion.isNetCustomDialogShow) {
                            showNetCustomDialog(versionEntity!!)
                        } else {
                            if (!UpdateVersion.isUpdateDownloadWithBrowser) {
                                httpClient?.downloadApk(versionEntity, DownloadCallback())
                            } else {
                                val i =
                                    Intent(Intent.ACTION_VIEW, Uri.parse(versionEntity?.url))
                                mContext?.startActivity(i)
                            }
                        }
                    }
                }
            } else {
                if (UpdateVersion.isHintVersion) {
                    Toast.makeText(
                        mContext,
                        if (mContext == null) "当前已是最新版" else mContext!!.getString(R.string.version_hint),
                        Toast.LENGTH_LONG
                    ).show()
                }
            }
        } else {
            if (UpdateVersion.isHintVersion) {
                Toast.makeText(
                    mContext,
                    if (mContext == null) "当前已是最新版" else mContext!!.getString(R.string.version_hint),
                    Toast.LENGTH_LONG
                ).show()
            }
        }
    }

    /**
     * 获取当前app版本
     */
    private val packageInfo: PackageInfo?
        get() {
            var pinfo: PackageInfo? = null
            if (mContext != null) {
                try {
                    pinfo = mContext?.packageManager?.getPackageInfo(
                        mContext!!.packageName, 0
                    )
                } catch (e: PackageManager.NameNotFoundException) {
                    e.printStackTrace()
                }
            }
            return pinfo
        }

    /**
     * 更新提示
     */
    @SuppressLint("WrongConstant")
    private fun showUpdateUICustom(versionEntity: VersionEntity) {
        val dialogFragment = UpdateDialogFragment()
        dialogFragment.setStyle(DialogFragment.STYLE_NO_TITLE, 0)
        dialogFragment.setTitle(
            String.format(
                if (mContext == null) "发现新版本%s" else mContext!!.getString(R.string.version_title),
                """
     
     ${versionEntity.versionNo}
     """
                    .trimIndent()
            )
        )
        dialogFragment.setContentTitleText(
            if (TextUtils.isEmpty(versionEntity.title)) if (mContext == null) "更新的内容" else mContext!!.getString(
                R.string.version_content_title
            ) else versionEntity.title
        )
        dialogFragment.setContentText(
            String.format(
                if (mContext == null) "%s" else mContext!!.getString(
                    R.string.version_content
                ), versionEntity.updateDesc
            )
        )
        dialogFragment.setNegativeBtn(
            if (mContext == null) "暂不" else mContext!!.getString(R.string.version_cancel),
            null
        )
        dialogFragment.setPositiveBtn(if (mContext == null) "立即更新" else mContext!!.getString(R.string.version_confirm),
            object : UpdateDialogFragment.OnClickListener {
                override fun onClick() {
                    val type = mContext.getConnectedType()
                    if (type != NetType.Wifi && UpdateVersion.isNetCustomDialogShow) {
                        showNetCustomDialog(versionEntity)
                    } else {
                        if (!UpdateVersion.isUpdateDownloadWithBrowser) {
                            httpClient?.downloadApk(versionEntity, DownloadCallback())
                        } else {
                            val i = Intent(Intent.ACTION_VIEW, Uri.parse(versionEntity.url))
                            mContext?.startActivity(i)
                        }
                    }
                }
            })
        try {
            dialogFragment.show(findActivity(mContext!!)!!.supportFragmentManager, "")
        } catch (e: IllegalStateException) {
            Log.e("UpdateDialogFragment", e.message!!)
        }
    }

    /**
     * 手机网络dialog
     */
    @SuppressLint("WrongConstant")
    private fun showNetCustomDialog(versionEntity: VersionEntity?) {
        val dialogFragment = UpdateDialogFragment()
        dialogFragment.setStyle(DialogFragment.STYLE_NO_TITLE, 0)
        dialogFragment.setContentText(
            if (mContext == null) "当前在无WIFI的情况下下载，请确定是否使用流量继续下载。" else mContext!!.getString(
                R.string.no_wifi_hint
            )
        )
        dialogFragment.setNetHint(true)
        dialogFragment.setNegativeBtn(
            if (mContext == null) "取消" else mContext!!.getString(R.string.no_wifi_hint_cancel),
            null
        )
        dialogFragment.setPositiveBtn(if (mContext == null) "继续下载" else mContext!!.getString(R.string.no_wifi_hint_confirm),
            object : UpdateDialogFragment.OnClickListener {
                override fun onClick() {
                    if (!UpdateVersion.isUpdateDownloadWithBrowser) {
                        httpClient?.downloadApk(versionEntity, DownloadCallback())
                    } else {
                        val i = Intent(Intent.ACTION_VIEW, Uri.parse(versionEntity?.url))
                        mContext?.startActivity(i)
                    }
                }
            })
        try {
            dialogFragment.show(findActivity(mContext!!)!!.supportFragmentManager, "")
        } catch (e: IllegalStateException) {
            Log.e("UpdateDialogFragment", e.message!!)
        }
    }

    fun findActivity(context: Context?): FragmentActivity? {
        if (context is FragmentActivity) {
            return context
        }
        return if (context is ContextWrapper) {
            val wrapper = context as ContextWrapper?
            findActivity(wrapper?.baseContext)
        } else {
            null
        }
    }

    companion object {
        @SuppressLint("StaticFieldLeak")
        private var mContext: Context? = null
        fun nativeAssertGet(URL: String?): InputStream? {
            var inputStream: InputStream? = null
            try {
                inputStream = mContext?.assets?.open(URL!!)
            } catch (e: Exception) {
                e.printStackTrace()
            }
            return inputStream
        }
    }
}